{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "453aa072",
   "metadata": {},
   "source": [
    "## Tutorial 2: Xenium datasets preprocess demo\n",
    "#### 1. Download the raw data from 10x Genoimcs ：Metadata:https://www.10xgenomics.com/datasets/human-liver-data-xenium-human-multi-tissue-and-cancer-panel-1-standard; Alignment file: https://cf.10xgenomics.com/samples/xenium/1.9.0/Xenium_V1_hLiver_nondiseased_section_FFPE/Xenium_V1_hLiver_nondiseased_section_FFPE_he_imagealignment.csv OME tif: https://cf.10xgenomics.com/samples/xenium/1.9.0/Xenium_V1_hLiver_nondiseased_section_FFPE/Xenium_V1_hLiver_nondiseased_section_FFPE_he_image.ome.tif\n",
    "#### 2. Preprocess-data format\n",
    "- `re_image.tif`：Raw histology image\n",
    "- `pseudo_st.csv`：Gene count matrix。\n",
    "  - Row 1: Gene names.\n",
    "  - Row 2 and after: Each row is a spot.\n",
    "    - Column 1: Spot ID.\n",
    "    - Column 2 and after: Each column is a gene.\n",
    "- `pseudo_locs.csv`：Spot locations\n",
    "  - Row 1: Header\n",
    "  - Row 2 and after: Each row is a spot. Must match rows in `pseudo_st.csv`\n",
    "    - Column 1: Spot ID\n",
    "    - Column 2: x-coordinate (horizontal axis). Must be in the same space as axis-1 (column) of the array indices of pixels in `re_image.tif`.\n",
    "    - Column 3: y-coordinate (vertical axis). Must be in the same space as axis-0 (row) of the array indices of pixels in `re_image.tif`.\n",
    "- `mask.png`：Tissue segmentation based on the valid sequencing area.\n",
    "- `gene_names.txt`: Gene list.\n",
    "- `pixel-size-raw.txt`：Side length (in micrometers) of pixels in `re_image.tif`.This value is usually between 0.1 and 1.0.\n",
    "- `radius-raw.txt`：Number of pixels per spot radius in `re_image.tif`.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1b5c5f8f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✓ Image aligned, padded, and saved.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import cv2\n",
    "import numpy as np\n",
    "os.environ[\"OPENCV_IO_MAX_IMAGE_PIXELS\"] = pow(2, 40).__str__()\n",
    "\n",
    "def pad_to_multiple_of_224(img, fill_value=255):\n",
    "    H, W = img.shape[:2]\n",
    "    new_H = ((H + 223) // 224) * 224\n",
    "    new_W = ((W + 223) // 224) * 224\n",
    "\n",
    "    pad_H = new_H - H\n",
    "    pad_W = new_W - W\n",
    "\n",
    "    img_padded = np.pad(\n",
    "        img,\n",
    "        ((0, pad_H), (0, pad_W), (0, 0)),\n",
    "        mode='constant',\n",
    "        constant_values=fill_value\n",
    "    )\n",
    "    return img_padded\n",
    "\n",
    "\n",
    "he = cv2.imread(\"/home/lixiaoyu/VISD/xenium-data/Raw-data/Xenium_V1_hLiver_nondiseased_section_FFPE_he_image.ome.tif\")\n",
    "ome = cv2.imread(\"/home/lixiaoyu/VISD/xenium-data/Raw-data/morphology.ome.tif\")\n",
    "M = pd.read_csv(\"/home/lixiaoyu/VISD/xenium-data/Raw-data/Xenium_V1_hLiver_nondiseased_section_FFPE_he_imagealignment.csv\", header=None).values\n",
    "\n",
    "\n",
    "scale = 0.2125 / 0.5\n",
    "new_width = int(he.shape[1] * scale)\n",
    "new_height = int(he.shape[0] * scale)\n",
    "he_small = cv2.resize(he, (new_width, new_height), interpolation=cv2.INTER_AREA)\n",
    "S = np.array([[scale, 0, 0], [0, scale, 0], [0, 0, 1]])\n",
    "M_new = M.copy()\n",
    "M_new[0, 2] *= scale  \n",
    "M_new[1, 2] *= scale  \n",
    "dst_width = int(ome.shape[1] * scale)\n",
    "dst_height = int(ome.shape[0] * scale)\n",
    "img = cv2.warpPerspective(he_small, M_new, dsize=(dst_width, dst_height))\n",
    "img = img.astype(np.uint8)\n",
    "img = pad_to_multiple_of_224(img)\n",
    "cv2.imwrite(\"/home/lixiaoyu/VISD/xenium-data/Pre-data/re_image.tif\", img)\n",
    "\n",
    "print(\"✓ Image aligned, padded, and saved.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8c25a3fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing cells: 100%|██████████| 239271/239271 [17:42<00:00, 225.18it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✓ Xenium 3D gene tensor has been saved to: /home/lixiaoyu/VISD/xenium-data/Pre-data/genes_3D.pkl\n"
     ]
    }
   ],
   "source": [
    "import cv2\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "from shapely.geometry import Polygon\n",
    "import pickle\n",
    "from skimage.draw import polygon2mask\n",
    "from tqdm import tqdm\n",
    "\n",
    "PATHS_Healthy = {\n",
    "    \"matrix_dir\": \"/home/lixiaoyu/VISD/xenium-data/Raw-data/cell_feature_matrix\",\n",
    "    \"boundaries_csv\": \"/home/lixiaoyu/VISD/xenium-data/Raw-data/cell_boundaries.csv\",\n",
    "    \"cells_csv\": \"/home/lixiaoyu/VISD/xenium-data/Raw-data/cells.csv\",\n",
    "    \"gene_names_txt\": \"/home/lixiaoyu/VISD/xenium-data/Pre-data/gene_names.txt\",\n",
    "    \"he_image\": \"/home/lixiaoyu/VISD/xenium-data/Pre-data/re_image.tif\",\n",
    "    \"output_pkl\": \"/home/lixiaoyu/VISD/xenium-data/Pre-data/genes_3D.pkl\"\n",
    "}\n",
    "\n",
    "\n",
    "def Cal_area_2poly(data1, data2):\n",
    "    poly1 = Polygon(data1).convex_hull\n",
    "    poly2 = Polygon(data2).convex_hull\n",
    "    if not poly1.intersects(poly2):\n",
    "        return 0\n",
    "    return poly1.intersection(poly2).area\n",
    "\n",
    "def load_dataset(paths):\n",
    "    adata = sc.read_10x_mtx(paths[\"matrix_dir\"], var_names='gene_symbols')\n",
    "    adata.var_names_make_unique()\n",
    "    gene_names = list(adata.var_names)\n",
    "    counts = np.array(adata.X.todense())\n",
    "\n",
    "    boundaries = pd.read_csv(paths[\"boundaries_csv\"])\n",
    "    cells = pd.read_csv(paths[\"cells_csv\"])\n",
    "\n",
    "    all_cell_ids = pd.concat([cells['cell_id'], boundaries['cell_id']]).unique()\n",
    "    cell_id_to_idx = {cell_id: idx for idx, cell_id in enumerate(all_cell_ids)}\n",
    "    cells['cell_id'] = cells['cell_id'].map(cell_id_to_idx)\n",
    "    boundaries['cell_id'] = boundaries['cell_id'].map(cell_id_to_idx)\n",
    "\n",
    "    boundaries['vertex_x'] *= 2\n",
    "    boundaries['vertex_y'] *= 2\n",
    "\n",
    "    image = cv2.imread(paths[\"he_image\"])\n",
    "    return adata, counts, gene_names, boundaries, cells, image\n",
    "\n",
    "def build_cnts(counts, boundaries, cells, image_shape, gene_indices, grid_size=16):\n",
    "    cnts = np.zeros((image_shape[0] // grid_size, image_shape[1] // grid_size, len(gene_indices)))\n",
    "    for c in tqdm(range(len(cells)), desc=\"Processing cells\"):\n",
    "        data1 = boundaries[boundaries['cell_id'] == c][['vertex_x', 'vertex_y']].values\n",
    "\n",
    "        pixes = (data1 // grid_size * grid_size).astype(int)\n",
    "        pixes = np.unique(pixes, axis=0)\n",
    "\n",
    "        min_x, min_y = np.min(data1, axis=0).astype(int) // grid_size * grid_size\n",
    "        max_x, max_y = np.max(data1, axis=0).astype(int) // grid_size * grid_size\n",
    "\n",
    "        all_pixes = np.array([[x, y] for x in range(min_x, max_x + grid_size, grid_size)\n",
    "                                      for y in range(min_y, max_y + grid_size, grid_size)])\n",
    "\n",
    "        mask = polygon2mask(\n",
    "            ((max_y - min_y) // grid_size + 1, (max_x - min_x) // grid_size + 1),\n",
    "            (data1 - [min_x, min_y]) / grid_size\n",
    "        )\n",
    "\n",
    "        valid_mask = mask[\n",
    "            (all_pixes[:, 1] // grid_size - min_y // grid_size),\n",
    "            (all_pixes[:, 0] // grid_size - min_x // grid_size)\n",
    "        ]\n",
    "        valid_pixes = all_pixes[valid_mask]\n",
    "        pixes = np.unique(np.vstack((pixes, valid_pixes)), axis=0)\n",
    "\n",
    "        areas = []\n",
    "        for p in pixes:\n",
    "            x, y = p\n",
    "            data2 = np.array([[x, y], [x, y+grid_size], [x+grid_size, y+grid_size], [x+grid_size, y]])\n",
    "            areas.append(Cal_area_2poly(data1, data2))\n",
    "\n",
    "        area1 = sum(areas)\n",
    "        for i, p in enumerate(pixes):\n",
    "            x, y = p\n",
    "            h, w = y // grid_size, x // grid_size\n",
    "            if h >= cnts.shape[0] or w >= cnts.shape[1]:\n",
    "                continue\n",
    "            cnts[h, w, :] += counts[c, gene_indices] * (areas[i] / area1)\n",
    "    return cnts\n",
    "\n",
    "adata_healthy, counts_healthy, genes_healthy, boundaries_healthy, cells_healthy, image_healthy = load_dataset(PATHS_Healthy)\n",
    "\n",
    "common_genes = sorted(set(genes_healthy))\n",
    "idx_healthy = [adata_healthy.var_names.get_loc(g) for g in common_genes]\n",
    "\n",
    "pd.Series(common_genes).to_csv(PATHS_Healthy[\"gene_names_txt\"], index=False, header=False)\n",
    "\n",
    "cnts_healthy = build_cnts(counts_healthy, boundaries_healthy, cells_healthy, image_healthy.shape, idx_healthy)\n",
    "with open(PATHS_Healthy[\"output_pkl\"], 'wb') as f:\n",
    "    pickle.dump(cnts_healthy, f)\n",
    "\n",
    "print(\"✓ Xenium 3D gene tensor has been saved to:\", PATHS_Healthy[\"output_pkl\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f14c22ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.ndimage import binary_dilation\n",
    "import numpy as np\n",
    "from matplotlib import pyplot as plt\n",
    "import pickle\n",
    "\n",
    "\n",
    "with open(\"/home/lixiaoyu/VISD/xenium-data/Pre-data/genes_3D.pkl\", 'rb') as f:\n",
    "    gene = pickle.load(f)\n",
    "    matrix = gene\n",
    "\n",
    "mask = np.any(matrix, axis=2)\n",
    "# mask[:70, :] = 0\n",
    "mask = mask.astype(np.uint8) * 255\n",
    "\n",
    "bool_mask = mask == 255\n",
    "dilated_mask = binary_dilation(bool_mask)\n",
    "new_mask = dilated_mask.astype(np.uint8) * 255\n",
    "\n",
    "plt.imsave(\"/home/lixiaoyu/VISD/xenium-data/Pre-data/mask.png\", mask, cmap='gray')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f1ee8779",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Image loaded from /home/lixiaoyu/VISD/xenium-data/Pre-data/mask.png\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import cv2\n",
    "import pandas as pd\n",
    "from utils import load_image\n",
    "\n",
    "distance = 100  \n",
    "pixel_size = 0.5\n",
    "he = cv2.imread(\"/home/lixiaoyu/VISD/xenium-data/Pre-data/re_image.tif\")\n",
    "list = []\n",
    "for i in range(0, int(he.shape[1] * pixel_size - 27.5), distance):\n",
    "    if i/100 % 2 == 0:\n",
    "        k = 0\n",
    "    else:\n",
    "        k = 50\n",
    "    for j in range(0, int(he.shape[0] * pixel_size - 27.5 - (127.5-k)), distance):\n",
    "        x = i + 27.5\n",
    "        y = j + 27.5 + k\n",
    "        list.append([x, y])\n",
    "list = np.array(list).astype(np.float32)\n",
    "list = list // pixel_size  \n",
    "\n",
    "df = pd.DataFrame(list, columns=['x', 'y'])\n",
    "\n",
    "\n",
    "mask = load_image(\"/home/lixiaoyu/VISD/xenium-data/Pre-data/mask.png\") > 0\n",
    "mask = mask[:, :, 0]\n",
    "\n",
    "rows_to_delete = []\n",
    "H, W = mask.shape\n",
    "for i in range(H):\n",
    "    for j in range(W):\n",
    "        if mask[i, j] == False:  \n",
    "            row_to_delete = df[(df['x'] // 16 - 3 == j) & (df['y'] // 16 - 3 == i)]\n",
    "            if not row_to_delete.empty:\n",
    "                rows_to_delete.append(row_to_delete.index[0])\n",
    "\n",
    "df = df.drop(rows_to_delete)\n",
    "df.to_csv(\"/home/lixiaoyu/VISD/xenium-data/Pre-data/pseudo_locs.csv\", index=False, float_format='%.2f')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f00e6d41",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dataframe loaded from /home/lixiaoyu/VISD/xenium-data/Pre-data/pseudo_locs.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Image.py (3368): Image size (366937088 pixels) exceeds limit of 300000000 pixels, could be decompression bomb DOS attack.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Image loaded from /home/lixiaoyu/VISD/xenium-data/Pre-data/re_image.tif\n",
      "[INFO] Skipped 0 patches due to boundary issues.\n",
      "[DONE] Saved pseudo_st.csv and pseudo_locs.csv.\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pickle\n",
    "import pandas as pd\n",
    "from PIL import Image\n",
    "Image.MAX_IMAGE_PIXELS = 300000000\n",
    "\n",
    "def load_image(filename, verbose=True):\n",
    "    img = Image.open(filename)\n",
    "    img = np.array(img)\n",
    "    if img.ndim == 3 and img.shape[-1] == 4:\n",
    "        img = img[..., :3]  # remove alpha channel\n",
    "    if verbose:\n",
    "        print(f'Image loaded from {filename}')\n",
    "    return img\n",
    "\n",
    "def load_tsv(filename, index=None):\n",
    "    index_col = 0 if index else None\n",
    "    df = pd.read_csv(filename, header=0, index_col=index_col)\n",
    "    print(f'Dataframe loaded from {filename}')\n",
    "    return df\n",
    "\n",
    "def get_locs(prefix, target_shape=None):\n",
    "    df = load_tsv(f'{prefix}pseudo_locs.csv')\n",
    "    orig_locs = np.stack([df['x'], df['y']], -1)\n",
    "\n",
    "    used_locs = orig_locs.copy()\n",
    "    if target_shape is not None:\n",
    "        wsi = load_image(f'{prefix}re_image.tif')\n",
    "        current_shape = np.array(wsi.shape[:2])\n",
    "        rescale_factor = current_shape // target_shape\n",
    "        used_locs = used_locs.astype(float)\n",
    "        used_locs /= rescale_factor\n",
    "        used_locs = used_locs.round().astype(int)\n",
    "\n",
    "    return orig_locs, used_locs\n",
    "\n",
    "def get_disk_mask(radius, boundary_width=None):\n",
    "    radius_ceil = int(np.ceil(radius))\n",
    "    locs = np.meshgrid(\n",
    "        np.arange(-radius_ceil, radius_ceil + 1),\n",
    "        np.arange(-radius_ceil, radius_ceil + 1),\n",
    "        indexing='ij')\n",
    "    locs = np.stack(locs, -1)\n",
    "    distsq = (locs**2).sum(-1)\n",
    "    isin = distsq <= radius**2\n",
    "    if boundary_width is not None:\n",
    "        isin &= distsq >= (radius - boundary_width)**2\n",
    "    return isin\n",
    "\n",
    "def get_patches_flat(img, used_locs, mask, orig_locs):\n",
    "    shape = np.array(mask.shape)\n",
    "    center = shape // 2\n",
    "    r = np.stack([-center, shape - center], -1)\n",
    "    x_list = []\n",
    "    valid_orig_locs = []\n",
    "    skipped = 0\n",
    "\n",
    "    for idx, s in enumerate(used_locs):\n",
    "        y1 = s[1] + r[0][0]\n",
    "        y2 = s[1] + r[0][1]\n",
    "        x1 = s[0] + r[1][0]\n",
    "        x2 = s[0] + r[1][1]\n",
    "\n",
    "        if y1 < 0 or y2 > img.shape[0] or x1 < 0 or x2 > img.shape[1]:\n",
    "            skipped += 1\n",
    "            continue\n",
    "\n",
    "        patch = img[y1:y2, x1:x2]\n",
    "\n",
    "        if patch.shape[:2] != mask.shape:\n",
    "            skipped += 1\n",
    "            continue\n",
    "\n",
    "        x = patch[mask]  # shape: (n_pixels, genes)\n",
    "        x_list.append(x)\n",
    "        valid_orig_locs.append(orig_locs[idx])\n",
    "\n",
    "    print(f\"[INFO] Skipped {skipped} patches due to boundary issues.\")\n",
    "    x_list = np.stack(x_list)\n",
    "    valid_orig_locs = np.array(valid_orig_locs)\n",
    "    return x_list, valid_orig_locs\n",
    "\n",
    "\n",
    "\n",
    "prefix = \"/home/lixiaoyu/VISD/xenium-data/Pre-data/\"\n",
    "\n",
    "with open(prefix + \"genes_3D.pkl\", 'rb') as f:\n",
    "    img = pickle.load(f)\n",
    "\n",
    "radius = 55 / 16\n",
    "mask = get_disk_mask(radius)\n",
    "\n",
    "\n",
    "orig_locs, used_locs = get_locs(prefix, np.array([img.shape[0], img.shape[1]]))\n",
    "\n",
    "x, filtered_orig_locs = get_patches_flat(img, used_locs, mask, orig_locs)\n",
    "\n",
    "cnts = np.sum(x, axis=1)\n",
    "\n",
    "\n",
    "with open(prefix + \"gene_names.txt\", 'r') as file:\n",
    "    gene_names = [line.strip() for line in file]\n",
    "\n",
    "assert cnts.shape[1] == len(gene_names), \"Mismatch between gene count and gene_names\"\n",
    "\n",
    "cnts_df = pd.DataFrame(data=cnts, columns=gene_names)\n",
    "cnts_df.to_csv(prefix + \"pseudo_st.csv\", index=False)\n",
    "\n",
    "filtered_df = pd.DataFrame(filtered_orig_locs, columns=[\"x\", \"y\"])\n",
    "filtered_df.to_csv(prefix + \"pseudo_locs.csv\", index=False)\n",
    "\n",
    "print(\"[DONE] Saved pseudo_st.csv and pseudo_locs.csv.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5afb87ff",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sr4v",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
